rs/sns/governance/src/governance/cast_vote_and_cascade_follow_tests.rs

```diff
@@ -245,3 +246,776 @@ fn test_cast_vote_and_cascade_follow_critical_vs_normal_proposals() {
+                                followees: vec![Followee { neuron_id: Some(nid(0)), alias: None }],
+                            },
+                        }
+                    }),
+                ),
+                neuron(
+                    nid(2),
+                    btreemap! {
+                        u64::from(&Action::DeregisterDappCanisters(Default::default())) => Followees {
+                            followees: vec![nid(0)],
+                        },
+                    },
+                    None,
+                ),
+            ],
+            Box::new(|directly_cast_vote| btreemap! {
+                nid(0).to_string() => Ballot {
+                    vote: directly_cast_vote as i32,
+                    voting_power,
+                    cast_timestamp_seconds,
+                },
+                nid(1).to_string() => Ballot {
+                    vote: directly_cast_vote as i32,
+                    voting_power,
+                    cast_timestamp_seconds,
+                },
+                nid(2).to_string() => Ballot {
+                    vote: directly_cast_vote as i32,
+                    voting_power,
+                    cast_timestamp_seconds,
+                },
+            }),
+        ),
+        (
+            "Three neurons:  N0 -- function --> N1; N0 -- function --> N2.",
+            Action::DeregisterDappCanisters(Default::default()),
+            Topic::CriticalDappOperations,
+            vec![
+                voting_neuron.clone(),
+                neuron(
+                    nid(1),
+                    btreemap! {
+                        u64::from(&Action::DeregisterDappCanisters(Default::default())) => Followees {
+                            followees: vec![nid(0)],
+                        },
+                    },
+                    None,
+                ),
+                neuron(
+                    nid(2),
+                    btreemap! {
+                        u64::from(&Action::DeregisterDappCanisters(Default::default())) => Followees {
+                            followees: vec![nid(0)],
+                        },
+                    },
+                    None,
+                ),
+            ],
+            Box::new(|directly_cast_vote| btreemap! {
+                nid(0).to_string() => Ballot {
+                    vote: directly_cast_vote as i32,
+                    voting_power,
+                    cast_timestamp_seconds,
+                },
+                nid(1).to_string() => Ballot {
+                    vote: directly_cast_vote as i32,
+                    voting_power,
+                    cast_timestamp_seconds,
+                },
+                nid(2).to_string() => Ballot {
+                    vote: directly_cast_vote as i32,
+                    voting_power,
+                    cast_timestamp_seconds,
+                },
+            }),
+        ),
+    ];
+
+    for (label, action, topic, neurons, expected_ballots) in test_cases {
```

Comment:
One nit - it is probably better to have a single test-runner function that each test calls, so that you get a list of failures.

When you do it this way, you can't really tell if your change broke all tests related or just a single specific case, which seems like less useful feedback.

I've come to the conclusion over the years of doing things very much like this that it's almost always better to have different top-level tests that call a common runner, rather than a list that's iterated.

Reply:
Completely agree, I just somehow forgot about this detail when I started writing these. I'll see if I can get to refactor these tests in a follow-up PR. I think the stuff added here is already going to improve the situation in terms of how precisely we can specify how liquid democracy works in the SNS. :)